Update documentation

Sidduppal opened the pull request:
:memo: Add step-by-step tutorial on how to run Autometa    
:fire: Remove Rest API
:art: Add docs/source/_build to .gitignore
:memo: Update autometa install guidelines. Added docker to it.
:memo: Add benchmarking page
:memo: Add Automappa to examining results
:art: Replaced shell with bash in parse_argparse.py
:memo: Add packages to install for developers in contributing guidelines"

jason-c-kwan commented on docs/source/running-autometa.rst:
```diff
@@ -2,85 +2,266 @@
+
+Use the following command to run the binning:
+
+.. code-block:: bash
+
+    # Archaeal binning
+    autometa-binning <path/to/kmers_norm.tsv> \
+    <path/to/coverage_table.tsv> <path/to/archaea.markers.tsv> \
+    <path/to/archaea_binning.tsv> --embedded-kmers <path/to/embedded_kmers.tsv> \
+    --taxonomy <path/to/taxonomy.tsv> --clustering-method <dbscan or hdbscan> --domain archaea
+
+    # Bacterial binning
+    autometa-binning <path/to/kmers_norm.tsv> \
+    <path/to/coverage_table.tsv> <path/to/bacterial.markers.tsv> \
+    <path/to/bacteria_binning.tsv> --embedded-kmers <path/to/embedded_kmers.tsv> \
+    --taxonomy <path/to/taxonomy.tsv> --clustering-method <dbscan or hdbscan> --domain bacteria
```
I initially couldn't run this (see Issue #133)

jason-c-kwan replied:
Evan commented on Slack and thinks the command is wrong:

```
Looks like you’ve provided the kmers table where you should have provided the markers table. The sacc column is from the markers table

The kmer embedding is loaded already, so if you’ve copied and pasted from the docs, then the docs has the input paths wrong. Otherwise, this is not a bug, the kmers table is loaded in and then the markers table is trying to be loaded in but fails as the incorrect file was provided.
```

jason-c-kwan replied:
Apparently the positional arguments have to be after the optional ones.

jason-c-kwan commented on docs/source/running-autometa.rst:
```diff
@@ -2,85 +2,266 @@
+
+Use the following command to run the unclustered recruitment step:
+
+.. code-block:: bash
+
+    # Archaea
+    autometa-unclustered-recruitment <path/to/kmers_norm.tsv> \
+    <path/to/coverage_table.tsv> <path/to/archaea_binning.tsv> \
+    <path/to/archaea.markers.tsv> <path/to/arachaea_unclustered_recruitment.tsv> \
+    --taxonomy <path/to/taxonomy.tsv> --classifier decision_tree
+
+    # Bacteria
+    autometa-unclustered-recruitment <path/to/kmers_norm.tsv> \
+    <path/to/coverage_table.tsv> <path/to/bacteria_binning.tsv> \
+    <path/to/bacteria.markers.tsv> <path/to/bacteria_unclustered_recruitment.tsv> \
+    --taxonomy <path/to/taxonomy.tsv> --classifier decision_tree
```
I didn't even get to trying this - I just didn't have much hope after the last five or so commands failed.

jason-c-kwan reviewed (changes requested):
I ran into so many problems here. My approach was to go through the documentation (at least "running autometa" and try out the commands). Details in comments. I would recommend writing it as a walk through using one of the simulated datasets as an example, then as you are writing it, actually do the commands yourself (with copy/paste). This will show any errors and also errors with the data files (which I think I uncovered some of). The user will be really disheartened if they can't even do the example in the documentation, which is how I felt frankly.

WiscEvan commented:
Just a note, pre-commit hooks should also be addressed in contributing. Not just `conda install pre-commit` but also running `pre-commit install` in the repo so they are ready.

WiscEvan commented on docs/source/contribution-guidelines.rst:
```diff
@@ -5,3 +5,19 @@ Contributing Guidelines
 "Autometa is an open-source project developed on
 GitHub. If you would like to help develop Autometa or
 have ideas for new features please see our `contributing guidelines <https://github.com/KwanLab/Autometa/blob/master/.github/CONTRIBUTING.md>`__
```
Note: We have two different versions of `CONTRIBUTING.md`. The `dev` branch refers to using `black` formatting whereas the master branch does not

WiscEvan commented on docs/source/contribution-guidelines.rst:
```diff
@@ -5,3 +5,19 @@ Contributing Guidelines
+
+.. code-block:: bash
+
+    conda install -n autometa -c conda-forge\
```
```suggestion
    conda install -n autometa -c conda-forge \
```

WiscEvan commented on docs/source/examining-results.rst:
```diff
@@ -2,20 +2,51 @@
+
+An interactive interface for exploration and refinement of metagenomes Automappa is a tool built to interface with Autometa output to help you explore your binning results.
+
+For details, see the Automappa page `here. <https://github.com/WiscEvan/Automappa>`__
```
```suggestion
For details, see the `Automappa page <https://github.com/WiscEvan/Automappa>`__
```

WiscEvan commented on docs/source/examining-results.rst:
```diff
@@ -2,20 +2,51 @@
+Binning summary
+===============
+
+Create a master table with all the binning annotations including embedded kmers, taxonomy and coverage.
+
+.. code-block:: python
+
+    import pandas as pd
+    kmers = pd.read_csv("kmers.am_clr.bhsne.tsv", sep='\t', index_col='contig')
+    binning = pd.read_csv("binning.hdbscan.tsv", sep='\t', index_col='contig')
+    binning = pd.read_csv("taxonomy.tsv", sep='\t', index_col='contig')
+    coverage=pd.read_csv("coverage.tsv", sep="\t", index_col="contig")
+    # how='right' will only take contigs in the binning table.
+    df = pd.merge(kmers, binning, left_index=True, right_index=True, how='right')
+    df = pd.merge(taxonomy, df, left_index=True, right_index=True, how='right')
+    df = pd.merge(coverage, df, left_index=True, right_index=True, how='right')
+    df.write_csv("master.tsv", sep="\t", index="contig")
```
I agree with Jason

WiscEvan commented on docs/source/examining-results.rst:
```diff
@@ -2,20 +2,51 @@
+    library("ggplot2")
+    fpath="master.tsv"
+    data = read.table(fpath, header=TRUE, sep='\t')
+    ggplot(data, aes(x=x, y=y, color=cluster, group=cluster)) + geom_point(size=(sqrt(data$length))/100, shape=20, alpha=0.5) + theme_classic() + xlab('BH-tSNE X') + ylab('BH-tSNE Y') 
```
Perhaps this should be formatted with line breaks?

WiscEvan commented on docs/source/examining-results.rst:
```diff
@@ -2,20 +2,51 @@
+
+    library("ggplot2")
+    fpath="master.tsv"
+    ggplot(data, aes(x=x, y=y, color=phylum, group=phylum)) + geom_point(size=(sqrt(data$length))/100, shape=20, alpha=0.5) + theme_classic() + xlab('BH-tSNE X') + ylab('BH-tSNE Y') 
```
Perhaps this should be formatted with line breaks?

WiscEvan commented on docs/source/index.rst:
```diff
@@ -11,9 +11,11 @@ Autometa documentation
-    python autometa.py metagenome.config
+.. code-block:: bash
+    
+    ./nextflow run autometa.nf
```
`./nextflow` assumes the user has downloaded a nextflow executable into that directory. 

If the user has installed nextflow via conda (should be able to with `conda install -n autometa --file=requirements.txt`) the command will be available from the conda env, _not_ as an executable in the user's directory

```suggestion
    nextflow run autometa.nf
```

WiscEvan commented on docs/source/install.rst:
```diff
@@ -6,52 +6,63 @@ Install
+==================
 
+#. Install miniconda_
+#. Create a new environment ``conda create -n autometa "python>=3.7"``
+#. Install autometa ``conda install -c conda-forge -c bioconda autometa --yes``
+#. Actiavate autometa envoirnonment ``conda activate autometa``
```
NOTE: We have requirements.txt in our root directory. So the environment can be created and installed with one command:

## Install from source (using make)

```bash
cd $HOME
git clone https://github.com/KwanLab/Autometa.git
cd $HOME/Autometa
# create autometa conda environment
make create_environment
# activate env
conda activate autometa
# install autometa source code
make install
```

## Install from source (full commands)

```bash
cd $HOME
git clone https://github.com/KwanLab/Autometa.git
cd $HOME/Autometa
# Construct the environment from the listed requirements.
conda create -n autometa --file=requirements.txt
# Install the autometa code base from source
python setup.py install
```

WiscEvan replied:
## build the docker image from source

Also note, you can building a docker image for your clone of the Autometa repo.

```bash
cd $HOME
git clone https://github.com/KwanLab/Autometa.git
cd $HOME/Autometa
# This will tag the image as jason-c-kwan/autometa:<your current branch>
make image
```

WiscEvan commented on docs/source/install.rst:
```diff
@@ -6,52 +6,63 @@ Install
+Docker image
+============
+
+You can also run Autometa using a prebuild Docker image. 
+
+#. Install Docker_
+#. Run the following commands
+
+    .. code-block:: bash
+
+        git clone https://github.com/KwanLab/Autometa.git
+        docker pull jasonkwan/autometa:latest
```
1. There is no need to clone the repository if the user just wants to use the docker image.
2. Also no need to run docker pull as docker recognize if an image is not present/available.

User should be able to just run 

```bash
docker run jason-c-kwan/autometa:dev
docker run jason-c-kwan/autometa:latest
... whatever other tags we plan on having available
```

and docker will retrieve the image if they do not already have it.

WiscEvan commented on docs/source/install.rst:
```diff
@@ -6,52 +6,63 @@ Install
-Now download autometa to your desired directory using:
+    # Check dependencies to see if Autometa environment is appropriately configured
+    python autometa.py --check-dependencies
 
-    ``git clone https://github.com/KwanLab/Autometa.git``
+    # If any of the checks return False, you can check which failed using
+    python autometa.py --check-dependencies --debug
```
:bug: These commands are no longer available

WiscEvan commented on docs/source/install.rst:
```diff
@@ -6,52 +6,63 @@ Install
+
+    # Install packages for testing
+    conda install -n autometa -c conda-forge \
+        black pre_commit pytest pytest-cov pytest-html pytest-repeat pytest-variables gdown --yes
```
Not sure the `--yes` should be here.

WiscEvan commented on docs/source/install.rst:
```diff
@@ -6,52 +6,63 @@ Install
+    # Download test data
+    gdown https://drive.google.com/uc\?\id=1bSlPldaq3C6Cf9Y5Rm7iwtUDcjxAaeEk -O tests/data/test_data.json
+
+Run the tests using ``make test``. This will run the tests and make sure that everything is working well. 
```
```suggestion
Run the tests using ``make unit_test``. This will run the tests and make sure that everything is working well. 
```

WiscEvan commented on docs/source/simulated_community.csv:
```diff
@@ -0,0 +1,9 @@
+﻿Community,Num. Genomes,Num. Control Sequences
```
📝 Add section on using these links. Or simply make some of these commands available to the user
See issue-#110 (assigned @ajlail98). May be worth a discussion b/w @Sidduppal and @ajlail98.

# Download simulated communities 

## Example for the 78Mbp community

```bash
gdown --id 15CB8rmQaHTGy7gWtZedfBJkrwr51bb2y
```
You can get the file ID by navigating to any of the files and right clicking, then selecting the `get link` option. This will have a `copy link` button that you should use.

The link will look like this: `https://drive.google.com/file/d/15CB8rmQaHTGy7gWtZedfBJkrwr51bb2y/view?usp=sharing`

The file ID is within the `/` forward slashes between `file/d/` and `/`, e.g.

```bash
# Pasted from copy link button:
https://drive.google.com/file/d/15CB8rmQaHTGy7gWtZedfBJkrwr51bb2y/view?usp=sharing
#                 begin file ID ^ ------------------------------^ end file ID
```

Now that we have the File ID, you can specify the ID or use the `drive.google.com` prefix

```bash
file_id="15CB8rmQaHTGy7gWtZedfBJkrwr51bb2y"
gdown https://drive.google.com/uc?id=${file_id} -O metagenome.fna.gz
# or
gdown --id 15CB8rmQaHTGy7gWtZedfBJkrwr51bb2y
```

Either should work

WiscEvan commented on docs/source/simulated_community.csv:
```diff
@@ -0,0 +1,9 @@
+`1250Mbp <https://drive.google.com/drive/folders/17hnBBfmmAmj7on5JzSxR73zGVwhMDzUG?usp=sharing>`__,319,"41,507"
+`2500Mbp <https://drive.google.com/drive/folders/1re57q-mwLLq_qzdFkdA2eJoh41uMeEvl?usp=sharing>`__,656,"67,702"
+`5000Mbp <https://drive.google.com/drive/folders/1gFN2jUdY9o2kYDoBzLEIyggkHv58bTG3?usp=sharing>`__,"1,288","140,529"
+`10000Mbp <https://drive.google.com/drive/folders/1YaEn6rQvBiXLgyIWkxc-EruiYAorCIes?usp=sharing>`__,"2,638","285,262"
```
```suggestion
`10000Mbp <https://drive.google.com/drive/folders/1YaEn6rQvBiXLgyIWkxc-EruiYAorCIes?usp=sharing>`__,"2,638","285,262"

```

WiscEvan reviewed (changes requested):
Haven't went through the entire documentation, but there are a number of commands that need to be updated. I've also placed some replies to help fill out some of the docs. Also, https://github.com/KwanLab/Autometa/pull/120#discussion_r595197563 should likely be added within the "unit tests" section

WiscEvan commented on docs/source/contribution-guidelines.rst:
```diff
@@ -5,3 +5,19 @@ Contributing Guidelines
+.. code-block:: bash
+
+    conda install -n autometa -c conda-forge\
+        black pre_commit pytest pytest-cov pytest-html pytest-repeat pytest-variables sphinx sphinx_rtd_theme
```
See [(comment)](https://github.com/KwanLab/Autometa/pull/120/files#r595331991) on #120

WiscEvan commented:
# Current Nextflow walkthrough

NOTE: These instructions are for working off of the [WiscEvan/nextflow  branch](https://github.com/WiscEvan/Autometa/tree/nextflow)

## Install Autometa (use nextflow without docker)

```bash
git clone https://github.com/WiscEvan/Autometa
cd Autometa
## Check out nextflow branch
git checkout nextflow
#List all make options
make
#Create conda env for Autometa
make create_environment
#Activate Autometa
conda activate autometa
#Install Autometa
make install
```

## Configure nextflow with your 'executor'

For nextflow to run the Autometa pipeline through a job scheduler you will need to update the respective 'profile' section in nextflow's config file. Each 'profile' may be configured with any available scheduler as noted in the [nextflow executors docs](https://www.nextflow.io/docs/latest/executor.html "Nextflow Executors documentation"). By default nextflow will use your local computer as the 'executor'. The next section briefly walks through nextflow executor configuration to run with the slurm job scheduler.

### SLURM

NOTE: You can find the available slurm partitions by running `sinfo`
Run sinfo to see what partitions are available (e.g. ours is depicted below)

<img width="624" alt="SLURM partition information output from the sinfo command" src="https://user-images.githubusercontent.com/25933122/111811927-f10fdb00-88ad-11eb-936f-11c1de9d0dfb.png">

Will need to change `$HOME/Autometa/nextflow.config`

```groovy
// Find this section of code in nextflow.config
  }
  cluster {
    process.executor = "slurm"
    // queue is the slurm partition to use.
    // Set SLURM partition with queue directive.
    process.queue = "queue" // <<-- change this to whatever your partition is called
    // See https://www.nextflow.io/docs/latest/executor.html#slurm for more details.
  }
```

More parameters that are available for the slurm executor are listed in the [nextflow executor docs for slurm](https://www.nextflow.io/docs/latest/executor.html#slurm)

### CHTC

```groovy
// Find this section of code in nextflow.config
  }
  chtc {
    process.executor = "condor"
    // See https://www.nextflow.io/docs/latest/executor.html#htcondor for more configuration options.
  }
```

More parameters that are available for the slurm executor are listed in the [nextflow executor docs for HTCondor](https://www.nextflow.io/docs/latest/executor.html#htcondor)

## Running nextflow with docker

Building our docker image (Assuming docker is installed)

```bash
# NOTE: this will be `make image` 
# but `make image` currenty would be jason-c-kwan/autometa:nextflow
# so just use the below command for the time being
docker build . -t jason-c-kwan/autometa:dev
```

## Set Autometa parameters

Can alter default template parameters config file here: `$HOME/Autometa/nextflow/parameters.config`

### Data Inputs

**NOTE: Data inputs must be wrapped in 'single quotes' or "double quotes"**

```groovy
params.metagenome = null // <<-- Change this to path to scaffolds /path/to/scaffolds.fasta
params.interim = null // <<-- Path to where you want interim results stored - will make directory to store intermediate results
params.processed = null //<<-- Path to where you want final results stored - will make directory to store final results
```

Database directory path:

NOTE: Directory path must contain:

- diamond formatted nr file => nr.dmnd
- Extracted files from tarball taxdump.tar.gz
- prot.accession2taxid.gz

```groovy
params.ncbi_database = "$HOME/Autometa/autometa/databases/ncbi"  // <<-- Update this path to folder with all NCBI databases
```

### Use of Wildcards to run multiple metagenomes at once

NOTE: Wildcard characters will only be interpreted when "double quotes" are used

### Runtime Parameters

```groovy
params.cpus = 2  // <<-- No. of CPUs each job uses
```

### Autometa Parameters

```groovy
// Metagenome Length filtering
params.length_cutoff = 3000 // <<-- Smallest contiog you want binned (3000 is default)
// Kmer counting/normalization/embedding
params.kmer_size = 5
params.kmer_norm_method = "am_clr" // choices: "am_clr", "clr", "ilr"
params.kmer_pca_dimensions = 50
params.kmer_embed_method = "bhsne" // choices: "sksne", "bhsne", "umap"
params.kmer_embed_dimensions = 2
// Binning parameters
params.kingdom = "bacteria" // choices: "bacteria", "archaea"
params.classification_kmer_pca_dimensions = 50
params.kmer_embed_method = "bhsne" // choices: "bhsne", "sksne", "umap"
params.clustering_method = "dbscan" // choices: "dbscan", "hdbscan"
params.binning_starting_rank = "superkingdom" // choices: "superkingdom", "phylum", "class", "order", "family", "genus", "species"
params.classification_method = "decision_tree" // choices: "decision_tree", "random_forest"
params.completeness = 20.0 // Will keep clusters over 20% complete
params.purity = 95.0 // Will keep clusters over 95% pure
params.cov_stddev_limit = 25.0 // Will keep clusters less than 25% coverage std.dev.
params.gc_stddev_limit = 5.0 // Will keep clusters less than 5% GC% std.dev.
```

### Final run command

NOTE: This is from within the Autometa directory. If you would like to run the workflow outside of the autometa directory, you will need to supply an additional configuration argument to nextflow that holds your executor configuration.

#### Inside of the Autometa directory

NOTE: nextflow will find the nextflow.config file from the current directory, so the executor configuration will be available by default.

```bash
nextflow run \
    # main logic of autometa workflow
    $HOME/Autometa/main.nf \
    -profile cluster \
    # parameters configuration
    -c $HOME/Autometa/nextflow/parameters.config \
    # working directory where nextflow intermediate/tmp dirs/files will be written
    -w $HOME/work
```

#### Outside of the Autometa directory

```bash
nextflow run $HOME/Autometa/main.nf \
    # executor configuration
    -c $HOME/Autometa/nextflow.config \
    # parameters configuration
    -c </path/to/your/parameters.config> \
    # working directory where nextflow intermediate/tmp dirs/files will be written
    -w </path/to/nextflow/work/directory> \
    # available profiles mentioned above are cluster, chtc and standard (default)
    -profile <profile to use>
```

jason-c-kwan commented:
@WiscEvan If you eventually add that comment to the docs, you may want to change the `CHTC` heading, to something like `HTCondor` because CHTC is only really relevant to UW. There may need to be some discussion of shared filespaces, as @chasemc was talking about on Slack.

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+- `prot.accession2taxid.gz <ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz>`_
+- *nodes.dmp*, *names.dmp* and *merged.dmp* from `taxdump tarball <ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz>`_ 
+
+If you are running autometa for the first time you'll have to download these databases. Use ``autometa-update-databases --update-ncbi``. This will download the databases to the default path. You can check the default paths using ``autometa-config --print``. If you need to change the default download directory you can use ``autometa-config --section databases --option ncbi --value <path/to/new/ncb ...
```
This is kind of making me think we should just have a default path be the current directory if the user has not yet configured their databases section.

I.e. Without configuration

```bash
autometa-update-databases --update-ncbi
```

creates databases and specified sub-directories

1. mkdir -p ./databases/ncbi
2. downloads and formats ncbi files to `./databases/ncbi`

Maybe we need to rethink? Also would `autometa-databases` be a better entrypoint name?

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+    // Change the number of CPUs you want each job to use
+    params.cpus = 2
+
+Additional autometa parameteres
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
```suggestion
Additional autometa parameters
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+Additional autometa parameteres
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You can also adjust other pipeline parameteres that ultimate control how the binning would be done.
```
```suggestion
You can also adjust other pipeline parameters that ultimately control how the binning is performed.
```

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+
+You can also adjust other pipeline parameteres that ultimate control how the binning would be done.
+
+*params.length_cutoff* : Smallest contiog you want binned (default is 3000bp)
```
```suggestion
*params.length_cutoff* : Smallest contig you want binned (default is 3000bp)
```

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+
+*params.kmer_size* : kmer size to use
+
+*params.kmer_norm_method* : Which normaization method to use. Choices are centered log-ratio transformation (clr), Autometa's clr implementation (am_clr) and isometric log-ratio transformation (default is am_clr)
```
```suggestion
*params.norm_method* : Which normalization method to use. See :ref: kmer advanced usage section for details
```

> Choices are centered log-ratio transformation (clr), Autometa's clr implementation (am_clr) and isometric log-ratio transformation (default is am_clr)

This should be described in the kmer advanced usage section then we can just link to that here

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+
+*params.kmer_norm_method* : Which normaization method to use. Choices are centered log-ratio transformation (clr), Autometa's clr implementation (am_clr) and isometric log-ratio transformation (default is am_clr)
+
+*params.kmer_pca_dimensions* : Number of dimensions to reduce the initial 512 dimension big k-mer matrix to (default is 50)
```
```suggestion
*params.pca_dimensions* : Number of dimensions of which to reduce the initial k-mer frequencies matrix (default is 50)
```

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+
+*params.kmer_pca_dimensions* : Number of dimensions to reduce the initial 512 dimension big k-mer matrix to (default is 50)
+
+*params.kmer_embed_method* :  kmer-embedding method. Choices are "sksne", "bhsne", "umap" (default is bhsne)
```
```suggestion
*params.embedding_method* :  Choices are "sksne", "bhsne", "umap" (default is bhsne) See :ref: kmer advanced section for details
```

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+
+*params.kmer_embed_method* :  kmer-embedding method. Choices are "sksne", "bhsne", "umap" (default is bhsne)
+
+*params.kmer_embed_dimensions* : Final dimensions you want to reduce the pca reduced kmer matrix to (default is 2).
```
```suggestion
*params.embedding_dimensions* : Final dimensions of the kmer frequencies matrix (default is 2). See :ref: kmer advanced usage section for details
```

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+
+*params.kmer_embed_dimensions* : Final dimensions you want to reduce the pca reduced kmer matrix to (default is 2).
+
+*params.kingdom* : Bins contigs belonging to which kingdom. Choices are "bacteria" and "archaea" (default is bacteria)
```
```suggestion
*params.kingdom* : Bin contigs belonging to this kingdom. Choices are "bacteria" and "archaea" (default is bacteria)
```

We could by default go through marker annotation and binning for all kingdoms. This way, this parameter would be removed.

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+
+*params.kingdom* : Bins contigs belonging to which kingdom. Choices are "bacteria" and "archaea" (default is bacteria)
+
+*params.clustering_method* : Cluster contigs using which clustering method. Choices are "dbscan" and "hdbscan" (default is "dbscan")
+
+*params.binning_starting_rank* : Which taxonomic rank to start the binning from. Choices are "superkingdom", "phylum", "class", "order", "family", "genus", "species" (default is "superkingdom")
```
These should also contain references to the binning advanced usage section

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+
+*params.binning_starting_rank* : Which taxonomic rank to start the binning from. Choices are "superkingdom", "phylum", "class", "order", "family", "genus", "species" (default is "superkingdom")
+
+*params.classification_method* : Which clustering method to use for unclustered recruitment step. Choices are "decision_tree" and "random_forest" (default is "decision_tree")
```
This should reference the unclustered-recruitment advanced usage section

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+
+*params.classification_method* : Which clustering method to use for unclustered recruitment step. Choices are "decision_tree" and "random_forest" (default is "decision_tree")
+
+*params.completeness* :  Minimum completeness needed to keep a cluster (default is atleast 20% complete)
+
+*params.purity* : Minimum purity needed to keep a cluster (default is atleast 95% pure)
+
+*params.cov_stddev_limit* : Which clusters to keep depending on the covergae std.dev (default is 25%)
+
+*params.gc_stddev_limit* : Which clusters to keep depending on the GC% std.dev (default is 5%)
```
These should all reference the binning advanced usage section

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+.. note::
+    1. Run the following commands directly in ``bash``. Nexflow will automatically submit jobs to SLURM or any other sheduling system.
+    2. You can use `tmux <https://github.com/tmux/tmux/wiki>`_ or `screen <https://www.gnu.org/software/screen/>`_ in case you want to exit the window or disconnect from the server.
+    3. The pipeline must be launched from a node where the ``condor_submit`` command is available, that is, in a common usage scenario, the cluster head node.
+    4. The HTCondor executor for Nextflow_ does not support at this time the HTCondor ability to transfer input/output data to the corresponding job computing node. Therefore the data needs to be made accessible to the computing nodes using a shared file system directory from where the Nextflow_ workflow has to be executed (or specified via the -w option).
```
These are only relevant to users that want to run with an HTCondor executor. This should be moved to its own section. I think this would be intimidating to a new user trying out nextflow

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+
+.. code-block:: bash
+
+    # Clone the autometa repository
+    git clone git@github.com:KwanLab/Autometa.git
+    # Navigate to the clone repository
+    cd $HOME
+    # Run nextflow
+    nextflow run main.nf
```
This does not make sense. Where did this `main.nf` file come from?

```suggestion
    # Clone the autometa repository
    git clone git@github.com:KwanLab/Autometa.git $HOME/Autometa
    # Run nextflow
    nextflow run $HOME/Autometa/main.nf
```

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+
+Make sure that you have Docker_ and nextflow_ installed. You can run autometa pipeline using ``nextflow run KwanLab/Autometa -c parameters.config``
+
+In case you want to tweak some of the scripts and modify the pipeline you can clone the repository and then run autometa using ``main.nf``. This is completely optional and would be useful only for development purposes.
```
> useful only for development purposes

What if the user wants to run on their own scheduling system? i.e. run with slurm? This does not seem developmental but requires some tweaking with either parameters.config or the nextflow.config in the repo

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+Without docker
+^^^^^^^^^^^^^^
+
+By default autometa's implementation using nextflow_ makes use of Docker_. You can diasble it in either your ``nextflow.config`` file or the ``parameters.config`` file.
```
```suggestion
By default autometa's implementation using nextflow_ makes use of Docker_. You can disable it in either your ``nextflow.config`` file or the ``parameters.config`` file.
```

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+    fixOwnership = true
+    }
+
+Editing ``parameteres.config``:
```
```suggestion
Editing ``parameters.config``:
```

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
 
-Once you have resolved the dependencies and your checks are returning ``True``, you
-can move on to running the pipeline.
+Set autometa parameteres
+------------------------
```
```suggestion
Set autometa parameters
-----------------------
```

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
 
-Data preparation
-================
+Before we start running the pipeline we first need to alter some of the default paprameters in ``parameters.config``. A pre-made template is available on our `Autometa GitHub repository <https://github.com/KwanLab/Autometa>`_ which you can edit and customize as per your needs. You can access the ``parameters.config`` file template from `here <https://github.com/WiscEvan/Autometa/blob/4b4e3c60e076 ...
```
```suggestion
Before we start running the pipeline we first need to alter some of the default parameters in ``parameters.config``. A pre-made template is available on our `Autometa GitHub repository <https://github.com/KwanLab/Autometa>`_ which you can edit and customize as per your needs. You can access the ``parameters.config`` file template from `here <https://github.com/WiscEvan/Autometa/blob/4b4e3c60e076706e28deae4ae4d45f26b5df7dee/nextflow/parameters.config>`_. Go ahead and copy the file to the location of your choice and open it in your favorite text editor (vim, nano, vscode, etc).
```

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+
+By default autometa's implementation using nextflow_ makes use of Docker_. You can diasble it in either your ``nextflow.config`` file or the ``parameters.config`` file.
+
+Editing ``nextflow.config``:
```
Where is this nextflow.config file coming from? So far there has been no mention of where to find this

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+
+.. code-block:: bash
+
+    // Find this section of code in parameteres.config
+    docker.enabled = false // override use of docker to use local Autometa installation
```
This section does not exist in parameters.config.

```suggestion
    // Add this section of code in parameters.config
    docker.enabled = false // override use of docker to use local Autometa installation
```

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+    // Find this section of code in parameteres.config
+    docker.enabled = false // override use of docker to use local Autometa installation
+
+Now install autometa using one of the three install methods specified in :ref:`Install` (Directly using conda or from source). After the install you can run autometa using ``nextflow run KwanLab/Autometa -c parameters.config``. Nextflow_ would use the entrypoints created during the install to run autometa. Make sure to actiavte your conda environment before running incase you have installed using ...
```
This is not at all clear. There are three different methods suggested for installing autometa, but you are telling the user to run autometa basically using the GitHub repo. Why would they specify the repo for nextflow if they have just went through the trouble of a local install?

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+Other useful options
+^^^^^^^^^^^^^^^^^^^^
+
+``-c`` : In case you have configured nextflow_ with your executor (see below) and have made other modifications on how to run nextflow_ using your ``nexflow.config`` file, you can specify that file using the ``-c`` flag
```
This flag is required since you must provide your parameters config file with `-c paramters.config`. Doesn't seem necessarily like an "other" option in this case.

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+
+``-c`` : In case you have configured nextflow_ with your executor (see below) and have made other modifications on how to run nextflow_ using your ``nexflow.config`` file, you can specify that file using the ``-c`` flag
+
+``-w`` : BY default nextflow_ uses the current directory to store all the temporary files (called as ``work`` directory). You can change that using the ``-w`` flag
```
```suggestion
``-w`` : By default nextflow_ will create a ``work`` directory in the current directory to store all temporary files and nextflow related files. You can change this work directory using the ``-w`` flag
```

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+
+``-w`` : BY default nextflow_ uses the current directory to store all the temporary files (called as ``work`` directory). You can change that using the ``-w`` flag
+
+``-p`` : You can specify the executor to use using ``-p`` flag. Rightnow the available profiles are cluster, HTCondor and startard (default).
```
```suggestion
``-profile`` : You can specify the profile to use using ``-profile`` flag. For details on profiles see :ref: Configuring Profiles section
```

>Right now the available profiles are cluster, HTCondor and startard (default).

This should link to section on user configuring their own profiles. These will not work out of the box the end user.

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+^^^^^^^^^^^^^^^^^^^^^
 
-Fortunately, Autometa can construct this table for you with: ``python -m autometa.commmon.coverage``
+One of the most powerful features of nextflow_ is resuming the workflow from the last checkpoint. If your pipeline was interrupted for some reason you can resume it from the lat checkpoint using the resume flag (``-resume``). Eg, ``nextflow run KwanLab/Autometa -c parameters.config -resume``
```
```suggestion
One of the most powerful features of nextflow_ is resuming the workflow from the last completed process. If your pipeline was interrupted for some reason you can resume it from the last completed process using the resume flag (``-resume``). Eg, ``nextflow run KwanLab/Autometa -c parameters.config -resume``
```

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
 
-An Autometa metagenome job submission file is provided in the `tests` directory.
-You may find it here: `Autometa/tests/metagenome.config <https://github.com/KwanLab/Autometa/blob/dev/tests/metagenome.config>`_.
+After running nextflow you can see the execution statistics of your autometa run, including the time taken, CPUs used, RAM used, etc separately for each process. Nextflow would generate a summary report, a timeline report and a trace report automatically for you in the ``pipeline`` directory. You can read more about these execution reports `here <https://www.nextflow.io/docs/latest/tracing.html#e ...
```
```suggestion
After running nextflow you can see the execution statistics of your autometa run, including the time taken, CPUs used, RAM used, etc separately for each process. Nextflow would generate a summary report, a timeline report and a trace report automatically for you in the ``pipeline_info`` directory. You can read more about these execution reports `here <https://www.nextflow.io/docs/latest/tracing.html#execution-report>`_. 
```

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
-for calculating contig coverage. Note, if you have assembled your metagenome
-with SPAdes, you can specify ``cov_from_spades = True`` in the ``[parameters]``
-section and *only* your metagenome assembly is required.
+You can also visualize the entire workflow ie. create the DAG from the written DOT file. Install `Graphviz <https://graphviz.org/>`_ and do ``dot -Tpng < pipeline_info/autometa-dot > autometa-dag.png`` to get the in the ``png`` format.
```
What is DAG? or DOT?

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
-Notice *multiple* configs can be passed to run binning runs. In the future, this will allow
-pan-genome aware algorithms to use all metagenomes found within a project config.
+.. image:: ../img/slurm_partitions.png
+    :alt: Slurm partitions
```
```suggestion
    :alt: Screen shot of ``sinfo`` output showing ``queue`` listed under partition  
```

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
 
-Simply pass in the ``-m`` flag when calling a script to signify to python you are
-running an Autometa *module*.
+.. code-block:: bash
```
```suggestion
.. code-block:: groovy
```

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+
+Editing ``parameteres.config``:
+
+.. code-block:: bash
```
```suggestion
.. code-block:: groovy
```

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+
+Editing ``nextflow.config``:
+
+.. code-block:: bash
```
```suggestion
.. code-block:: groovy
```

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+This allows you to run the pipeline using the HTCondor resource manager. To do this you'll need to enable the HTCondor executor to condor value in the ``nextflow.config``.
 
-    from autometa.common.external import samtools
+.. code-block:: bash
```
```suggestion
.. code-block:: groovy
```

WiscEvan commented on docs/source/running-autometa.rst:
```diff
@@ -2,108 +2,271 @@
+        // See https://www.nextflow.io/docs/latest/executor.html#htcondor for more configuration options.
+    }
 
+More parameters that are available for the slurm executor are listed in the nextflow executor `docs for HTCondor <https://www.nextflow.io/docs/latest/executor.html#htcondor>`_.
```
```suggestion
More parameters that are available for the htcondor executor are listed in the nextflow executor `docs for HTCondor <https://www.nextflow.io/docs/latest/executor.html#htcondor>`_.
```

WiscEvan commented on docs/source/step-by-step-tutorial.rst:
```diff
@@ -0,0 +1,689 @@
+
+If you noticed I stored the ``78mbp_metagenome.bacteria.kmers.embedded.tsv`` file in the ``processed`` directory as no further analysis is required on the final.
+
+..note::
```
```suggestion
.. note::
```

WiscEvan reviewed (changes requested):
Wow, you have put a ton of work into this. Wonderful start and I think we are getting quite close to being ready for `dev` and a 2.0 release. Most of my comments are in the step-by-step tutorial section. I think it would be useful to add some "advanced usage" sections for the more involved scripts (binning, kmers, unclustered-recruitment). A few minor formatting details. I think it would also be helpful to have a nextflow tutorial that corresponds to the step-by-step so a user can follow along and familiarize themselves with both approaches. I had a few questions regarding some entrypoints and output files that I can change if you think it would be more appropriate. Most of the comments should be easy fixes. 👍  Once again, well done 🎊 👷 🔧

chasemc commented:
We had talked about my reformatting the nextflow code to more closely follow nf-core's guidelines (to be able to use their linter, etc.) and organizing the nextflow code to allow organized addition of more modules in the future (ie NSF aims).  I've been waiting until there was a working Nextflow branch to do all that, so maybe we should hold off on writing the Nextflow documentation until then?

Sidduppal commented:
> We had talked about my reformatting the nextflow code to more closely follow nf-core's guidelines (to be able to use their linter, etc.) and organizing the nextflow code to allow organized addition of more modules in the future (ie NSF aims). I've been waiting until there was a working Nextflow branch to do all that, so maybe we should hold off on writing the Nextflow documentation until then?

That sounds like a nice idea. I can start addressing comments which does not concern nextflow until @chasemc commits his changes.

Sidduppal commented on docs/source/step-by-step-tutorial.rst:
```diff
@@ -0,0 +1,689 @@
+
+Let us disect the above commands:
+
++---------------+----------------------------------------------------------------------------------------------+
+| Flag          | Function                                                                                     |
++===============+==============================================================================================+
+| --assembly    | Path to length filtered metagenome assembly                                                  |
++---------------+----------------------------------------------------------------------------------------------+
+| --from-spades | If the input assembly is generated using SPades then extract k-mer coverages from contig IDs |
++---------------+----------------------------------------------------------------------------------------------+
+| --bed         | Path to pre-prepared bed file                                                                |
++---------------+----------------------------------------------------------------------------------------------+
+| --bam         | Path to pre-prepared bam file                                                                |
++---------------+----------------------------------------------------------------------------------------------+
+| --lengths     | Path to table having length of each contig                                                   |
++---------------+----------------------------------------------------------------------------------------------+
+| --sam         | Path to pre-prepared sam file                                                                |
++---------------+----------------------------------------------------------------------------------------------+
+| --fwd-reads   | Path to forward reads                                                                        |
++---------------+----------------------------------------------------------------------------------------------+
+| --rev-reads   | Path to reverse reads                                                                        |
++---------------+----------------------------------------------------------------------------------------------+
+| --cpus        | Number of CPUs to use (default is to use all available CPUs)                                 |
++---------------+----------------------------------------------------------------------------------------------+
+| --out         | Path to coverage table of each contig                                                        |
++---------------+----------------------------------------------------------------------------------------------+
```
I think it would be hard to make a column denoting required/optional arguments as that would depend upon the type of files the user has.

Sidduppal commented on docs/source/step-by-step-tutorial.rst:
```diff
@@ -0,0 +1,689 @@
++------------------+-------------------------------------------------------------------------------------------------+
+| --classifier     | classifier to use for recruitment of contigs. Choices decision_tree (default) and random_forest |
++------------------+-------------------------------------------------------------------------------------------------+
+| --seed           | Seed to use for RandomState when initializing classifiers                                       |
```
I don't think I understand what you mean. Additional information is needed.

chasemc commented on docs/source/running-autometa.rst:
```diff
@@ -2,85 +2,268 @@
+
+*params.completeness* :  Minimum completeness needed to keep a cluster (default is atleast 20% complete). See :ref:`advanced-usage-binning` section for deails
+
+*params.purity* : Minimum purity needed to keep a cluster (default is atleast 95% pure). See :ref:`advanced-usage-binning` section for deails
```
"atleast"
"at least"

chasemc commented on docs/source/running-autometa.rst:
```diff
@@ -2,85 +2,268 @@
+
+*params.classification_method* : Which clustering method to use for unclustered recruitment step. Choices are "decision_tree" and "random_forest" (default is "decision_tree"). See :ref:`advanced-usage-unclustered-recruitment` section for deails
+
+*params.completeness* :  Minimum completeness needed to keep a cluster (default is atleast 20% complete). See :ref:`advanced-usage-binning` section for deails
```
"atleast"
"at least"

chasemc commented:
Will edit this comment later, putting this here for now:
Tuning hardware resources 
https://nf-co.re/usage/configuration#tuning-workflow-resources

chasemc commented:
@Sidduppal @WiscEvan @jason-c-kwan 

Can this be merged? And the nextflow documentaion be added in a new, smaller PR?

@Sidduppal I know you mentioned just writing the Nextflow stuff in a comment here and that you'd transfer it, but that requires a lot of duplicated effort between us and is probably harder for me than just writing in the .rst itself, especially since some of the file content needs to stay, some go.

jason-c-kwan commented:
If @Sidduppal changed this PR to "ready to review', I can go through it and add my comments and maybe approve to merge.

Sidduppal commented:
I have updated the PR however, I think it'll be better to review it after I have incorporated changes mention in PR#157 as the documentation would change after that merges.
